File: CURRENT/java/src/org/mariuszgromada/math/mxparser/parsertokens/CalculusOperator.java
/*
 * @(#)CalculusOperator.java        5.0.0    2022-03-20
 *
 * You may use this software under the condition of "Simplified BSD License"
 *
 * Copyright 2010-2022 MARIUSZ GROMADA. All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without modification, are
 * permitted provided that the following conditions are met:
 *
 *    1. Redistributions of source code must retain the above copyright notice, this list of
 *       conditions and the following disclaimer.
 *
 *    2. Redistributions in binary form must reproduce the above copyright notice, this list
 *       of conditions and the following disclaimer in the documentation and/or other materials
 *       provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY <MARIUSZ GROMADA> ``AS IS'' AND ANY EXPRESS OR IMPLIED
 * WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND
 * FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL <COPYRIGHT HOLDER> OR
 * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
 * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
 * ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
 * NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF
 * ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 *
 * The views and conclusions contained in the software and documentation are those of the
 * authors and should not be interpreted as representing official policies, either expressed
 * or implied, of MARIUSZ GROMADA.
 *
 * If you have any questions/bugs feel free to contact:
 *
 *     Mariusz Gromada
 *     mariuszgromada.org@gmail.com
 *     http://mathparser.org
 *     http://mathspace.pl
 *     http://janetsudoku.mariuszgromada.org
 *     http://github.com/mariuszgromada/MathParser.org-mXparser
 *     http://mariuszgromada.github.io/MathParser.org-mXparser
 *     http://mxparser.sourceforge.net
 *     http://bitbucket.org/mariuszgromada/mxparser
 *     http://mxparser.codeplex.com
 *     http://github.com/mariuszgromada/Janet-Sudoku
 *     http://janetsudoku.codeplex.com
 *     http://sourceforge.net/projects/janetsudoku
 *     http://bitbucket.org/mariuszgromada/janet-sudoku
 *     http://github.com/mariuszgromada/MathParser.org-mXparser
 *     http://scalarmath.org/
 *     https://play.google.com/store/apps/details?id=org.mathparser.scalar.lite
 *     https://play.google.com/store/apps/details?id=org.mathparser.scalar.pro
 *
 *                              Asked if he believes in one God, a mathematician answered:
 *                              "Yes, up to isomorphism."
 */
package org.mariuszgromada.math.mxparser.parsertokens;

import org.mariuszgromada.math.mxparser.mXparser;
/**
 * Calculus Operators - mXparser tokens definition.
 *
 * @author         <b>Mariusz Gromada</b><br>
 *                 <a href="mailto:mariuszgromada.org@gmail.com">mariuszgromada.org@gmail.com</a><br>
 *                 <a href="http://mathspace.pl" target="_blank">MathSpace.pl</a><br>
 *                 <a href="http://mathparser.org" target="_blank">MathParser.org - mXparser project page</a><br>
 *                 <a href="http://github.com/mariuszgromada/MathParser.org-mXparser" target="_blank">mXparser on GitHub</a><br>
 *                 <a href="http://mxparser.sourceforge.net" target="_blank">mXparser on SourceForge</a><br>
 *                 <a href="http://bitbucket.org/mariuszgromada/mxparser" target="_blank">mXparser on Bitbucket</a><br>
 *                 <a href="http://mxparser.codeplex.com" target="_blank">mXparser on CodePlex</a><br>
 *                 <a href="http://janetsudoku.mariuszgromada.org" target="_blank">Janet Sudoku - project web page</a><br>
 *                 <a href="http://github.com/mariuszgromada/Janet-Sudoku" target="_blank">Janet Sudoku on GitHub</a><br>
 *                 <a href="http://janetsudoku.codeplex.com" target="_blank">Janet Sudoku on CodePlex</a><br>
 *                 <a href="http://sourceforge.net/projects/janetsudoku" target="_blank">Janet Sudoku on SourceForge</a><br>
 *                 <a href="http://bitbucket.org/mariuszgromada/janet-sudoku" target="_blank">Janet Sudoku on BitBucket</a><br>
 *                 <a href="https://play.google.com/store/apps/details?id=org.mathparser.scalar.lite" target="_blank">Scalar Free</a><br>
 *                 <a href="https://play.google.com/store/apps/details?id=org.mathparser.scalar.pro" target="_blank">Scalar Pro</a><br>
 *                 <a href="http://scalarmath.org/" target="_blank">ScalarMath.org</a><br>
 *
 * @version        5.0.0
 */
public final class CalculusOperator {
	/*
	 * CalculusOperator - token type id.
	 */
	public static final int TYPE_ID 					= 8;
	public static final String TYPE_DESC				= "Calculus Operator";
	/*
	 * CalculusOperator - tokens id.
	 */
	public static final int SUM_ID 						= 1;
	public static final int PROD_ID 					= 3;
	public static final int INT_ID 						= 5;
	public static final int DER_ID 						= 6;
	public static final int DER_LEFT_ID 				= 7;
	public static final int DER_RIGHT_ID				= 8;
	public static final int DERN_ID						= 9;
	public static final int FORW_DIFF_ID				= 10;
	public static final int BACKW_DIFF_ID				= 11;
	public static final int AVG_ID						= 12;
	public static final int VAR_ID						= 13;
	public static final int STD_ID						= 14;
	public static final int MIN_ID						= 15;
	public static final int MAX_ID						= 16;
	public static final int SOLVE_ID					= 17;
	/*
	 * CalculusOperator - tokens key words.
	 */
	public static final String SUM_STR					= "sum";
	public static final String SUM_STR_UNI_1			= "∑";
	public static final String SUM_STR_UNI_2			= "Σ";
	public static final String PROD_STR					= "prod";
	public static final String PROD_STR_UNI_1			= "∏";
	public static final String PROD_STR_UNI_2			= "ℿ";
	public static final String PROD_STR_UNI_3			= "Π";
	public static final String INT_STR					= "int";
	public static final String INT_STR_UNI_1			= "∫";
	public static final String DER_STR					= "der";
	public static final String DER_STR_UNI_1			= "∂";
	public static final String DER_LEFT_STR				= "der-";
	public static final String DER_LEFT_STR_UNI_1		= "∂-";
	public static final String DER_RIGHT_STR			= "der+";
	public static final String DER_RIGHT_STR_UNI_1		= "∂+";
	public static final String DERN_STR					= "dern";
	public static final String FORW_DIFF_STR			= "diff";
	public static final String FORW_DIFF_STR_UNI_1		= "∆";
	public static final String FORW_DIFF_STR_UNI_2		= "Δ";
	public static final String BACKW_DIFF_STR			= "difb";
	public static final String BACKW_DIFF_STR_UNI_1		= "∇";
	public static final String AVG_STR					= "avg";
	public static final String VAR_STR					= "vari";
	public static final String STD_STR					= "stdi";
	public static final String MIN_STR					= "mini";
	public static final String MAX_STR					= "maxi";
	public static final String SOLVE_STR				= "solve";
	/*
	 * CalculusOperator - syntax.
	 */
	public static final String SUM_SYN					= SyntaxStringBuilder.calculusIFromToExprOptBy(SUM_STR);
	public static final String SUM_SYN_UNI_1			= SyntaxStringBuilder.calculusIFromToExprOptBy(SUM_STR_UNI_1);
	public static final String SUM_SYN_UNI_2			= SyntaxStringBuilder.calculusIFromToExprOptBy(SUM_STR_UNI_2);
	public static final String PROD_SYN					= SyntaxStringBuilder.calculusIFromToExprOptBy(PROD_STR);
	public static final String PROD_SYN_UNI_1			= SyntaxStringBuilder.calculusIFromToExprOptBy(PROD_STR_UNI_1);
	public static final String PROD_SYN_UNI_2			= SyntaxStringBuilder.calculusIFromToExprOptBy(PROD_STR_UNI_2);
	public static final String PROD_SYN_UNI_3			= SyntaxStringBuilder.calculusIFromToExprOptBy(PROD_STR_UNI_3);
	public static final String INT_SYN					= SyntaxStringBuilder.calculusExprArgFromTo(INT_STR);
	public static final String INT_SYN_UNI_1			= SyntaxStringBuilder.calculusExprArgFromTo(INT_STR_UNI_1);
	public static final String DER_SYN					= SyntaxStringBuilder.calculusExprArgOptPoint(DER_STR);
	public static final String DER_SYN_UNI_1			= SyntaxStringBuilder.calculusExprArgOptPoint(DER_STR_UNI_1);
	public static final String DER_LEFT_SYN				= SyntaxStringBuilder.calculusExprArgOptPoint(DER_LEFT_STR);
	public static final String DER_LEFT_SYN_UNI_1		= SyntaxStringBuilder.calculusExprArgOptPoint(DER_LEFT_STR_UNI_1);
	public static final String DER_RIGHT_SYN			= SyntaxStringBuilder.calculusExprArgOptPoint(DER_RIGHT_STR);
	public static final String DER_RIGHT_SYN_UNI_1		= SyntaxStringBuilder.calculusExprArgOptPoint(DER_RIGHT_STR_UNI_1);
	public static final String DERN_SYN					= SyntaxStringBuilder.exprNArg(DERN_STR);
	public static final String FORW_DIFF_SYN			= SyntaxStringBuilder.calculusExprArgOptDelta(FORW_DIFF_STR);
	public static final String FORW_DIFF_SYN_UNI_1		= SyntaxStringBuilder.calculusExprArgOptDelta(FORW_DIFF_STR_UNI_1);
	public static final String FORW_DIFF_SYN_UNI_2		= SyntaxStringBuilder.calculusExprArgOptDelta(FORW_DIFF_STR_UNI_2);
	public static final String BACKW_DIFF_SYN			= SyntaxStringBuilder.calculusExprArgOptDelta(BACKW_DIFF_STR);
	public static final String BACKW_DIFF_SYN_UNI_1		= SyntaxStringBuilder.calculusExprArgOptDelta(BACKW_DIFF_STR_UNI_1);
	public static final String AVG_SYN					= SyntaxStringBuilder.calculusIFromToExprOptBy(AVG_STR);
	public static final String VAR_SYN					= SyntaxStringBuilder.calculusIFromToExprOptBy(VAR_STR);
	public static final String STD_SYN					= SyntaxStringBuilder.calculusIFromToExprOptBy(STD_STR);
	public static final String MIN_SYN					= SyntaxStringBuilder.calculusIFromToExprOptBy(MIN_STR);
	public static final String MAX_SYN					= SyntaxStringBuilder.calculusIFromToExprOptBy(MAX_STR);
	public static final String SOLVE_SYN				= SyntaxStringBuilder.calculusExprArgFromTo(SOLVE_STR);
	/*
	 * CalculusOperator - tokens description.
	 */
	public static final String SUM_DESC					= "Summation operator - SIGMA";
	public static final String PROD_DESC				= "Product operator - PI";
	public static final String INT_DESC					= "Definite integral operator";
	public static final String DER_DESC					= "Derivative operator";
	public static final String DER_LEFT_DESC			= "Left derivative operator";
	public static final String DER_RIGHT_DESC			= "Right derivative operator";
	public static final String DERN_DESC				= "n-th derivative operator";
	public static final String FORW_DIFF_DESC			= "Forward difference operator";
	public static final String BACKW_DIFF_DESC			= "Backward difference operator";
	public static final String AVG_DESC					= "Average operator";
	public static final String VAR_DESC					= "Bias-corrected sample variance operator";
	public static final String STD_DESC					= "Bias-corrected sample standard deviation operator";
	public static final String MIN_DESC					= "Minimum value";
	public static final String MAX_DESC					= "Maximum value";
	public static final String SOLVE_DESC				= "f(x) = 0 equation solving, function root finding";
	/*
	 * CalculusOperator - since.
	 */
	public static final String SUM_SINCE				= mXparser.NAMEv10;
	public static final String SUM_SINCE_UNI_1			= mXparser.NAMEv50;
	public static final String SUM_SINCE_UNI_2			= mXparser.NAMEv50;
	public static final String PROD_SINCE				= mXparser.NAMEv10;
	public static final String PROD_SINCE_UNI_1			= mXparser.NAMEv50;
	public static final String PROD_SINCE_UNI_2			= mXparser.NAMEv50;
	public static final String PROD_SINCE_UNI_3			= mXparser.NAMEv50;
	public static final String INT_SINCE				= mXparser.NAMEv10;
	public static final String INT_SINCE_UNI_1			= mXparser.NAMEv50;
	public static final String DER_SINCE				= mXparser.NAMEv10;
	public static final String DER_SINCE_UNI_1			= mXparser.NAMEv50;
	public static final String DER_LEFT_SINCE			= mXparser.NAMEv10;
	public static final String DER_LEFT_SINCE_UNI_1		= mXparser.NAMEv50;
	public static final String DER_RIGHT_SINCE			= mXparser.NAMEv10;
	public static final String DER_RIGHT_SINCE_UNI_1	= mXparser.NAMEv50;
	public static final String DERN_SINCE				= mXparser.NAMEv10;
	public static final String FORW_DIFF_SINCE			= mXparser.NAMEv10;
	public static final String FORW_DIFF_SINCE_UNI_1	= mXparser.NAMEv50;
	public static final String FORW_DIFF_SINCE_UNI_2	= mXparser.NAMEv50;
	public static final String BACKW_DIFF_SINCE			= mXparser.NAMEv10;
	public static final String BACKW_DIFF_SINCE_UNI_1	= mXparser.NAMEv50;
	public static final String AVG_SINCE				= mXparser.NAMEv24;
	public static final String VAR_SINCE				= mXparser.NAMEv24;
	public static final String STD_SINCE				= mXparser.NAMEv24;
	public static final String MIN_SINCE				= mXparser.NAMEv24;
	public static final String MAX_SINCE				= mXparser.NAMEv24;
	public static final String SOLVE_SINCE				= mXparser.NAMEv40;
}

File: CURRENT/typescript/org/mariuszgromada/math/mxparser/parsertokens/CalculusOperator.ts
import { mXparserConstants } from '../mXparserConstants';

/**
 * Calculus Operators - mXparser tokens definition.
 * 
 * @author         <b>Mariusz Gromada</b><br>
 * <a href="mailto:mariuszgromada.org@gmail.com">mariuszgromada.org@gmail.com</a><br>
 * <a href="http://github.com/mariuszgromada/MathParser.org-mXparser" target="_blank">mXparser on GitHub</a><br>
 * 
 * @version        4.3.0
 * @class
 */
export class CalculusOperator {
    public static TYPE_ID: number = 8;

    public static TYPE_DESC: string = "Calculus Operator";

    public static SUM_ID: number = 1;

    public static PROD_ID: number = 3;

    public static INT_ID: number = 5;

    public static DER_ID: number = 6;

    public static DER_LEFT_ID: number = 7;

    public static DER_RIGHT_ID: number = 8;

    public static DERN_ID: number = 9;

    public static FORW_DIFF_ID: number = 10;

    public static BACKW_DIFF_ID: number = 11;

    public static AVG_ID: number = 12;

    public static VAR_ID: number = 13;

    public static STD_ID: number = 14;

    public static MIN_ID: number = 15;

    public static MAX_ID: number = 16;

    public static SOLVE_ID: number = 17;

    public static SUM_STR: string = "sum";

    public static PROD_STR: string = "prod";

    public static INT_STR: string = "int";

    public static DER_STR: string = "der";

    public static DER_LEFT_STR: string = "der-";

    public static DER_RIGHT_STR: string = "der+";

    public static DERN_STR: string = "dern";

    public static FORW_DIFF_STR: string = "diff";

    public static BACKW_DIFF_STR: string = "difb";

    public static AVG_STR: string = "avg";

    public static VAR_STR: string = "vari";

    public static STD_STR: string = "stdi";

    public static MIN_STR: string = "mini";

    public static MAX_STR: string = "maxi";

    public static SOLVE_STR: string = "solve";

    public static SUM_SYN: string = "sum( i, from, to, expr , <by> )";

    public static PROD_SYN: string = "prod( i, from, to, expr , <by> )";

    public static INT_SYN: string = "int( expr, arg, from, to )";

    public static DER_SYN: string = "der( expr, arg, <point> )";

    public static DER_LEFT_SYN: string = "der-( expr, arg, <point> )";

    public static DER_RIGHT_SYN: string = "der+( expr, arg, <point> )";

    public static DERN_SYN: string = "dern( expr, n, arg )";

    public static FORW_DIFF_SYN: string = "diff( expr, arg, <delta> )";

    public static BACKW_DIFF_SYN: string = "difb( expr, arg, <delta> )";

    public static AVG_SYN: string = "avg( i, from, to, expr , <by> )";

    public static VAR_SYN: string = "vari( i, from, to, expr , <by> )";

    public static STD_SYN: string = "stdi( i, from, to, expr , <by> )";

    public static MIN_SYN: string = "mini( i, from, to, expr , <by> )";

    public static MAX_SYN: string = "maxi( i, from, to, expr , <by> )";

    public static SOLVE_SYN: string = "solve( expr, arg, from, to )";

    public static SUM_DESC: string = "Summation operator - SIGMA";

    public static PROD_DESC: string = "Product operator - PI";

    public static INT_DESC: string = "Definite integral operator";

    public static DER_DESC: string = "Derivative operator";

    public static DER_LEFT_DESC: string = "Left derivative operator";

    public static DER_RIGHT_DESC: string = "Right derivative operator";

    public static DERN_DESC: string = "n-th derivative operator";

    public static FORW_DIFF_DESC: string = "Forward difference operator";

    public static BACKW_DIFF_DESC: string = "Backward difference operator";

    public static AVG_DESC: string = "Average operator";

    public static VAR_DESC: string = "Bias-corrected sample variance operator";

    public static STD_DESC: string = "Bias-corrected sample standard deviation operator";

    public static MIN_DESC: string = "Minimum value";

    public static MAX_DESC: string = "Maximum value";

    public static SOLVE_DESC: string = "f(x) = 0 equation solving, function root finding";

    public static SUM_SINCE: string; public static SUM_SINCE_$LI$(): string { if (CalculusOperator.SUM_SINCE == null) { CalculusOperator.SUM_SINCE = mXparserConstants.NAMEv10; }  return CalculusOperator.SUM_SINCE; }

    public static PROD_SINCE: string; public static PROD_SINCE_$LI$(): string { if (CalculusOperator.PROD_SINCE == null) { CalculusOperator.PROD_SINCE = mXparserConstants.NAMEv10; }  return CalculusOperator.PROD_SINCE; }

    public static INT_SINCE: string; public static INT_SINCE_$LI$(): string { if (CalculusOperator.INT_SINCE == null) { CalculusOperator.INT_SINCE = mXparserConstants.NAMEv10; }  return CalculusOperator.INT_SINCE; }

    public static DER_SINCE: string; public static DER_SINCE_$LI$(): string { if (CalculusOperator.DER_SINCE == null) { CalculusOperator.DER_SINCE = mXparserConstants.NAMEv10; }  return CalculusOperator.DER_SINCE; }

    public static DER_LEFT_SINCE: string; public static DER_LEFT_SINCE_$LI$(): string { if (CalculusOperator.DER_LEFT_SINCE == null) { CalculusOperator.DER_LEFT_SINCE = mXparserConstants.NAMEv10; }  return CalculusOperator.DER_LEFT_SINCE; }

    public static DER_RIGHT_SINCE: string; public static DER_RIGHT_SINCE_$LI$(): string { if (CalculusOperator.DER_RIGHT_SINCE == null) { CalculusOperator.DER_RIGHT_SINCE = mXparserConstants.NAMEv10; }  return CalculusOperator.DER_RIGHT_SINCE; }

    public static DERN_SINCE: string; public static DERN_SINCE_$LI$(): string { if (CalculusOperator.DERN_SINCE == null) { CalculusOperator.DERN_SINCE = mXparserConstants.NAMEv10; }  return CalculusOperator.DERN_SINCE; }

    public static FORW_DIFF_SINCE: string; public static FORW_DIFF_SINCE_$LI$(): string { if (CalculusOperator.FORW_DIFF_SINCE == null) { CalculusOperator.FORW_DIFF_SINCE = mXparserConstants.NAMEv10; }  return CalculusOperator.FORW_DIFF_SINCE; }

    public static BACKW_DIFF_SINCE: string; public static BACKW_DIFF_SINCE_$LI$(): string { if (CalculusOperator.BACKW_DIFF_SINCE == null) { CalculusOperator.BACKW_DIFF_SINCE = mXparserConstants.NAMEv10; }  return CalculusOperator.BACKW_DIFF_SINCE; }

    public static AVG_SINCE: string; public static AVG_SINCE_$LI$(): string { if (CalculusOperator.AVG_SINCE == null) { CalculusOperator.AVG_SINCE = mXparserConstants.NAMEv24; }  return CalculusOperator.AVG_SINCE; }

    public static VAR_SINCE: string; public static VAR_SINCE_$LI$(): string { if (CalculusOperator.VAR_SINCE == null) { CalculusOperator.VAR_SINCE = mXparserConstants.NAMEv24; }  return CalculusOperator.VAR_SINCE; }

    public static STD_SINCE: string; public static STD_SINCE_$LI$(): string { if (CalculusOperator.STD_SINCE == null) { CalculusOperator.STD_SINCE = mXparserConstants.NAMEv24; }  return CalculusOperator.STD_SINCE; }

    public static MIN_SINCE: string; public static MIN_SINCE_$LI$(): string { if (CalculusOperator.MIN_SINCE == null) { CalculusOperator.MIN_SINCE = mXparserConstants.NAMEv24; }  return CalculusOperator.MIN_SINCE; }

    public static MAX_SINCE: string; public static MAX_SINCE_$LI$(): string { if (CalculusOperator.MAX_SINCE == null) { CalculusOperator.MAX_SINCE = mXparserConstants.NAMEv24; }  return CalculusOperator.MAX_SINCE; }

    public static SOLVE_SINCE: string; public static SOLVE_SINCE_$LI$(): string { if (CalculusOperator.SOLVE_SINCE == null) { CalculusOperator.SOLVE_SINCE = mXparserConstants.NAMEv40; }  return CalculusOperator.SOLVE_SINCE; }
}
CalculusOperator["__class"] = "org.mariuszgromada.math.mxparser.parsertokens.CalculusOperator";